File: Work/LeetCode/Problems/254-FactorCombinations/FactorCombinations.java
/*************************************************************************
    > File Name: FactorCombinations.java
    > Author: lwei2
    > Mail: lwei2@IT.com 
    > The Description :
 ************************************************************************/

public class Solution
{
	public List<List<Integer>> getFactors(int n)
	{
		List<List<Integer>> result = new ArrayList<List<Integer>>();
		helper(result, new ArrayList<Integer>(), n, 2);
		return result;
	}

	public void helper(List<List<Integer>> result, List<Integer> item, int n, int start)
	{
		if(n <= 1)
		{
			if(item.size() > 1)
				result.add(new ArrayList<Integer>(item));
			return;
		}
		for(int i = start, i*i <= n; ++i)
		{
			if(n%i == 0)
			{
				item.add(i);
				helper(result, item, n/i, i);
				item.remove(item.size() - 1);				
			}
		}
		int i = n;
		item.add(i);
		helper(result, item, 1, i);
		item.remove(item.size() - 1);
	}
}

File: Work/LeetCode/Problems/254-FactorCombinations/FactorCombinations.cpp
/*************************************************************************
    > File Name: FactorCombinations.cpp
    > Author: lwei2
    > Mail: lwei2@IT.com 
    > The Description :
 ************************************************************************/


#include <iostream>
#include <vector>

using namespace std;

class Solution
{
	public:
		vector<vector<int> > getFactors1(int	n)
		{
			vector<vector<int> > res;
			vector<int> line;
			dfs(n,  2, res, line);
			return res;
		}
	private:
		void dfs(int n, int factor, vector<vector<int> >&res, vector<int> &line)
		{
			if(factor *factor > n)
				return;
			dfs(n, factor + 1, res, line);
			if(n%factor == 0)
			{
				line.push_back(factor);
				line.push_back(n/factor);
				res.push_back(line);
				line.pop_back();
				dfs(n/factor, factor, res, line);
				line.pop_back();
			}
		}

	public:
		vector<vector<int>> getFactors2(int n)
		{
			vector<vector<int>> result;
			vector<int> factors;
			getResult(n,&result, &factors);
			return result;
		}
		void getResult(const int n, vector<vector<int>> *result, vector<int> *factors)
		{
			for(int i = factors->empty() ? 2 : factors->back(); i <= n/i; ++i)
			{
				if(n%i == 0)
				{
					factors->emplace_back(i);
					factors->emplace_back(n/i);
					result->emplace_back(*factors);
					factors->pop_back();
					getResult(n/i, result, factors);
					factors->pop_back();
				}
			}
		}

};



int main(void)
{
	return 0;
}
